locksmith/src/websub/helpers/renewKey.ts

```diff
@@ -0,0 +1,130 @@
+import { ethers, Wallet, Contract, constants } from 'ethers'
+import { networks } from '@unlock-protocol/networks'
+import { NetworkConfig } from '@unlock-protocol/types'
+import { KeyRenewal } from '../../models'
+import { purchaserCredentials } from '../../../config/config'
+
+interface RenewKeyParams {
+  keyId: number
+  lockAddress: string
+  network: number
+}
+
+const abi = [
+  {
+    inputs: [
+      {
+        internalType: 'uint256',
+        name: '_tokenId',
+        type: 'uint256',
+      },
+      {
+        internalType: 'address',
+        name: '_referrer',
+        type: 'address',
+      },
+    ],
+    name: 'renewMembershipFor',
+    outputs: [],
+    stateMutability: 'nonpayable',
+    type: 'function',
+  },
+  {
+    inputs: [],
+    name: 'gasRefundValue',
+    outputs: [
+      {
+        internalType: 'uint256',
+        name: '_refundValue',
+        type: 'uint256',
+      },
+    ],
+    stateMutability: 'view',
+    type: 'function',
+  },
+  {
+    inputs: [],
+    name: 'publicLockVersion',
+    outputs: [
+      {
+        internalType: 'uint16',
+        name: '',
+        type: 'uint16',
+      },
+    ],
+    stateMutability: 'pure',
+    type: 'function',
+  },
+]
+
+export const renewMembershipFor = async (
+  network: number,
+  lock: any,
+  keyId: number,
+  signer?: any
+) => {
+  const renewalInfo = {
+    network,
+    keyId,
+    lockAddress: lock.address,
+  }
+
+  // make sure reccuring payments are supported
+  if ((await lock.publicLockVersion()) < 10) {
+    return {
+      ...renewalInfo,
+      msg: 'Renewal only supported for lock v10+',
+    }
+  }
+
+  // We only renew keys for which we are getting a gas refund whose value cover the gas spent.
+  const gasRefund = await lock.gasRefundValue()
+
+  // estimate gas for the renewMembership function.
+  const gasLimit = await lock.estimateGas.renewMembershipFor(
+    keyId,
+    constants.AddressZero
+  )
+  if (gasRefund.lte(gasLimit)) {
```

Comment:
That won't work because `gasLimit` is in "native currency" (ether on mainnet, matic on polygon...) vs `gasRefund` is in ERC20 tokens.

We need to do a conversion to $ of these 2.

Generally I think we should isolate that logic in its own function `isWorthRenewing(lock, keyId)` so that we can also "overwrite" its behavior. For example in the early days we should be fine as long as we don't spend more than say $0.03

Reply:
yep added